dns/psl.py

```diff
@@ -0,0 +1,194 @@
+        entities as part of their namespace.
+
+        *allow_unlisted_gtlds*, a ``bool``, with a default of ``True``.  If ``True``,
+        then the root node is a wildcard node, and gTLDs not listed in the
+        public suffix database will still be considered as public suffixes.  For
+        example, a query of "www.example.bogus-gtld." would have a public suffix
+        of "bogus-gtld." and a base domain of "example.bogus-gtld.".  If ``False``,
+        then ``None`` will be returned for gTLDs which are not listed.
+
+        *download_if_needed*: a ``bool``, defaulting to ``True``.  If ``True``, then
+        download the list from the *url* if *filename* does not exist.  If ``False``,
+        then *filename* must exist.
+
+        *url*: a ``str``.  The URL to use if downloading the public suffix list is
+        required; the default is the standard URL recommended by publicsuffix.org.
+        """
+
+        if download_if_needed:
+            if not _have_httpx:
+                raise ValueError(
+                    "download_if_needed is True but httpx is not available"
+                )
+            if not os.path.isfile(filename):
+                response = httpx.request("GET", url)
+                if response.status_code == 200:
+                    with open(filename, "w") as f:
+                        f.write(response.text)
+        self.suffixes = dns.namedict.NameDict()
+        if allow_unlisted_gtlds:
+            root_node: _Node = _WildNode(dns.name.root)
+        else:
+            root_node = _Node(dns.name.root)
+        self.suffixes[dns.name.root] = root_node
+        if categories is None:
+            categories = {"ICANN", "PRIVATE"}
+        assert categories is not None  # for mypy
+        pattern = re.compile("// ===(BEGIN|END) ([A-Z]+) DOMAINS===")
+        skipping = True
+        with open(filename, "r") as f:
+            self.mtime = os.fstat(f.fileno()).st_mtime
+            for l in f.readlines():
+                l = l.rstrip()
+                if l.startswith("//"):
+                    match = pattern.match(l)
+                    if match:
+                        op = match.group(1)
+                        category = match.group(2)
+                        skipping = not (category in categories and op == "BEGIN")
+                    continue
+                if l == "" or skipping:
+                    continue
+                if l.startswith("!"):
+                    exception = True
+                    l = l[1:]
+                else:
+                    exception = False
+                n = dns.name.from_text(l)
+                if n.is_wild():
+                    n = n.parent()  # remove leading "*" label
+                    node: _Node = _WildNode(n)
+                elif exception:
+                    node = _ExceptionNode(n)
+                else:
+                    node = _ExactNode(n)
+                if self.suffixes.has_key(n):
```

Comment:
Probably should be `n in self.suffixes`, as `has_key` usage has been deprecated for years (assuming that works; `NameDict` probably should be defining `__contains__`, but `MutableMapping` does).

Also, is it possible that some alternate PSL might contain the root domain, which would conflict with the automatically added root domain earlier?

Reply:
The root is not a public suffix, and the [PSL algorithm](https://github.com/publicsuffix/list/wiki/Format#algorithm) stops at the top level (with all TLDs considered public suffixes).

My interpretation of that is that the root is outside of the reach of the PSL algorithm, and its appearance in this PR's code is just a data structure convenience.

Reply:
I think that depends on whether the intent here is to support loading the official public suffix list, or other lists which are compatible.  If the latter, there is a possibility that it could contain the root, as I don't see anything in the spec that explicitly forbids it.

I don't see anything in the linked algorithm that prevents there being a rule at the root.  Also, according to the algorithm:
```
If no rules match, the prevailing rule is "*".
```
So, maybe the PR should be adding a default wildcard rule.

Reply:
> If the latter, there is a possibility that it could contain the root, as I don't see anything in the spec that explicitly forbids it.

But there's no point to that, because all TLDs are public suffixes by definition (see your other quote above), so you can never climb up to the root as a TLD public suffix will always be in the way.

The code here handles that with the `allow_unlisted_gtlds` parameter which defaults to `True`.

Reply:
I'm not sure what you mean.  If there's a rule at `.`, then there will never be the case that no rules match, because that rule will always match at the end.  So, the text I quoted above would be irrelevant in this case.

I'm also not sure why any of this really matters.  I simply suggested avoiding a misleading error.  There's nothing in the spec that prevents a rule for `.`, so having a rule for `.` should not result in an error.  Even if you think it should result in an error, it should not be a redefinition error, as it's not a redefinition.